aten/src/ATen/core/Generator.h

```diff
@@ -1,24 +1,103 @@
 #pragma once
 
-#include <ATen/core/ATenGeneral.h>
 #include <stdint.h>
+#include <memory>
+#include <mutex>
+#include <deque>
+#include <atomic>
+#include <typeinfo>
+#include <utility>
+
+#include <c10/util/Exception.h>
+#include <c10/util/C++17.h>
+#include <ATen/Device.h>
+
+/**
+ * Note [Generator]
+ * ~~~~~~~~~~~~~~~~
+ * A Pseudo Random Number Generator (PRNG) is an engine that uses an algorithm to
+ * generate a seemingly random sequence of numbers, that may be later be used in creating
+ * a random distribution. Such an engine almost always maintains a state and requires a
+ * seed to start off the creation of random numbers. Often times, users have
+ * found it beneficial to be able to explicitly create, retain, and destroy
+ * PRNG states and also be able to have control over the seed value.
+ *
+ * A Generator in ATen gives users the ability to read, write and modify a PRNG engine.
+ * For instance, it does so by letting users seed a PRNG engine, fork the state of the
+ * engine, etc.
+ *
+ * By default, there is one generator per device, and a device's generator is
+ * lazily created. A user can use the torch.Generator() api to create their own generator.
+ * Currently torch.Generator() can only create a CPUGenerator.
+ */
+
+/**
+ * Note [Thread-safety and Generators]
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Generator and its derived classes are NOT thread-safe. Please use the public mutex_ when 
+ * using any methods from these classes. You can learn about the usage by looking into the 
+ * unittests (aten/src/ATen/cpu_generator_test.cpp). Note that we call
+ * std::lock_guard<std::mutex> AFTER every call of at::check_generator_with_default. This is
+ * because, check_generator_with_default uses the mutex_ in its definition and if you call
+ * std::lock_guard before check_generator_with_default, you'll hang inside 
+ * check_generator_with_default.
```

Comment:
You can eliminate this footgun by making it a `recursive_mutex`. Not sure if you *should* do it, but it would prevent this failure mode.

Reply:
The recursive_mutex is not needed anymore since I'm getting rid of the lock_guard in check_generator_with_default. Also deleting this part of the note.